oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resourceserver/authentication/JwtAccessTokenVerifier.java

```diff
@@ -67,23 +74,33 @@ public AccessTokenJwtVerifier(JwtVerifier jwtVerifier, Duration maxClockSkew) {
 	}
 
 	@Override
-	public void verifyClaims(Jwt jwt) throws JwtException {
-		Instant expiry = jwt.getExpiresAt();
+	public void verify(Map<String, Object> tokenAttributes) throws OAuth2AuthenticationException {
+		JwtClaimAccessor jwtClaimAccessor = () -> tokenAttributes;
+
+		Instant expiry = jwtClaimAccessor.getExpiresAt();
 
 		if ( expiry != null ) {
 			if ( Instant.now().isAfter(expiry.plus(maxClockSkew)) ) {
-				throw new JwtException(String.format("Jwt expired at {}", jwt.getExpiresAt()));
```

Comment:
I'm not sure I agree that using Jwt -> using OAuth2. For example, there are folks who use JWT to express their csrf tokens. I realize csrf is out-of-scope here, but the point I make is that I think it is a bit chancy for us to presume that `OAuth2AuthenticationException` is what an individual who is trying to use a `JwtVerifier` would intend.

I understand that we can easily say "Spring Security's support for JWT is intentionally  coupled to its support for OAuth2"--is that the intent?

Reply:
> the point I make is that I think it is a bit chancy for us to presume that OAuth2AuthenticationException is what an individual who is trying to use a JwtVerifier would intend

`JwtVerifier` (renamed to `OAuth2TokenVerifier`) would only be leveraged by an `AuthenticationProvider`, for example, `JwtAccessTokenAuthenticationProvider` and therefore should throw `AuthenticationException`. `JwtException` does not extend `AuthenticationException` instead it extends `RuntimeException`. Additionally, we will need to provide token verification for opaque token support so throwing `OAuth2AuthenticationException` works well rather than `JwtException`.

This interface should be general enough to suit both implementations - Jwt and Opaque (Introspection)

```
public interface OAuth2TokenVerifier {

	void verify(Map<String, Object> tokenAttributes) throws OAuth2AuthenticationException;

}
```

> I understand that we can easily say "Spring Security's support for JWT is intentionally coupled to its support for OAuth2"--is that the intent?

Correct, that is the intent that @rwinch and I discussed a while back. Spring Security's JWT/JOSE support is coupled to it's OAuth2/OIDC support. We do not intend it to be used as a standalone library.

Reply:
Again, @jgrandja and I had an extended off-line conversation about this, and so I'll summarize the high points here.

`spring-security-crypto` is an example of a library that is not directly related to authentication and authorization and yet we have it. There, likewise, may be an argument one day for a dedicated JWT library; however, insofar as our focus is on OAuth2, we want to make sure that we aren't declaring to the community that we are ready to both support JWT *and* OAuth2 with the 5.1 release. Really, Spring Security has no official plans to support JWT by itself, and so let's not do that inadvertently.